### DISCUSSION TITLE
como subir el reto

### DISCUSSION THREAD

**Discussion by RichardBecerra** [nodo-eafit-frontend/frontend-workshop-2024-1#13] [General]
Profe buenas tardes, tengo dudas de como tengo que subir el reto, la verdad estoy perdido con ese tema

**Comment by DanteCely** [comment]
Hola @RichardBecerra 
Ahí te recomiendo que veas los videos de la clase 30 y 31 de Mayo.
Otra cosa es que revises las instrucciones en el README.md del Repo.

Me confiras en qu eparte quedaste.

**Comment by EAlvarez90** [comment]
hola @DanteCely profe, he estado tratando de subir el proyecto, pero me aparece este error al seguir los pasos para el `commit` me ayudas para entender en que estoy fallando. Gracias 

<img width="953" alt="image" src="https://github.com/nodo-eafit-frontend/frontend-workshop-2024-f04/assets/170678278/00c8bb8f-b2bd-4262-9477-5b5d5c36f7ff">

**Reply by DanteCely** [reply]
Hola @EAlvarez90!
Lo que estoy observando es que estás ubicada en un commit específico. Debes estar ubicada en tu rama de trabajo para poder subir tus cambios. Tra de usar `git checkout <nombre-de-tu-rama>` para ubicarte a tu rama y poder subir tus cambios.

**Comment by kanjimon1** [comment]
Actualmente estoy trabajando en un formulario para el registro de horas extras, es un crud, pero manejo el insert, el update y el delete en modales, pero al hacer clic en el boton de crear, me abre el modal donde voy a consulta la infromación de un empleado, y me retorna la correctamente la información del employee id que ingresé pero, debido a los estados de los modales, como que se re-renderiza y no pinta la infomación y hace como que se refrecara el formulario parent; he intentado ya varias formas, pero ninguna me ha funcionado, consulté varios videos en youtube, foros y nada, sigue lo mismo; varias opciones de chagpt, claude y gemini pero casi todos llegan a las misma conslusiones; tengo código documentado porque es todas las formas que he intentado pero igual me sigue pasanodo y ya no tengo más opciones; adjunto la evidencia para su validación: 

![image](https://github.com/user-attachments/assets/31cc03bd-c456-4029-bd88-4dd34f3b0610)

![image](https://github.com/user-attachments/assets/a7766b65-6b5a-4af0-9271-ec4c0b161e20)

![image](https://github.com/user-attachments/assets/308ed6d4-e71c-4bef-bfe5-b9bcaa6a0428)

Clase parent dond están los modales de crear, actualizar y eliminar: 

`import React, { useState, useEffect, useContext, useCallback } from 'react';
import UserService from '../service/UserService';
//import background from '../../assets/images/mainBackground.png';
import { AuthContext } from '../context/AuthContext';
import { Outlet, Link, useNavigate, Form } from 'react-router-dom';
import CreateModal from './CreateExtraHourModal';
import UpdateModal from './UpdateExtraHoursModal';
import DeleteModal from './DeleteExtraHoursModal';
//import EmployeeSearchModal from './EmployeeSearchModela';

const UsersExtraHours = () => {

    const [extraHours, setExtraHours] = useState([]);
    const [loading, setLoading] = useState(true);
    const [error, setError] = useState(null);
    const { isAuthenticated, auth, logout } = useContext(AuthContext);
    const navigate = useNavigate();
    const [createModalOpen, setCreateModalOpen] = useState(false);
    const [updateModalOpen, setUpdateModalOpen] = useState(false);
    //const [selectedHour, setSelectedHour] = useState(null);
    const [selectedHour, setSelectedHour] = useState({});
    const [deleteModalOpen, setDeleteModalOpen] = useState(false);

    useEffect(() => {

        //const FetchExtraHoursUsersByUser = async (email) => {
        const fetchExtraHoursUsersByUser = async () => {
            try {
                const data = await UserService.getAllExtraHoursUsersByUser(auth.token, auth.email);
                console.log("RESPUESTA DESDE EL SERVER: ", data);
                // Map the response to the structure expected by extraHours
                const formattedData = data.map(hour => ({
                    id: hour.employeeId,
                    employee: {
                        employeeId: hour.employeeId,
                        employeeName: hour.employeeName,
                        salary: hour.salary,
                        area: { areaName: hour.areaName },
                        job: { jobName: hour.jobName }
                    },
                    amountExtraHours: hour.amountExtraHours,
                    startDatetime: hour.startDatetime,
                    endDatetime: hour.endDatetime,
                    totalExtraHour: hour.totalExtraHour,
                    totalPayment: hour.totalPayment,
                    extraHourType: {
                        description: hour.extraHourTypeDescription
                    }
                }));
                setExtraHours(formattedData);
                setLoading(false);
            } catch (err) {
                console.error('Error fetching user extra hours:', err);
                setError('Error loading user extra hours data');
                setLoading(false);
            }
        };


        if (isAuthenticated && auth.role === 'USER' && auth.email) {
            console.log("Auth state:", { isAuthenticated, role: auth.role, email: auth.email });
            fetchExtraHoursUsersByUser();
        }

        //}, [isAuthenticated, auth]);
        //useEffect(() => {

        console.log("ENTRA AL USE EFFECT");
        const fetchExtraHoursUsers = async () => {
            console.log("ENTRA AL USE EFFECT FECTH ADMIN");
            try {
                const data = await UserService.getAllExtraHoursUsers(auth.token);
                setExtraHours(data);
                setLoading(false);
            } catch (err) {
                console.error('Error fetching extra hours:', err);
                setError('Error loading extra hours data');
                setLoading(false);
            }
        };

        if (isAuthenticated && auth.role === 'ADMIN') {
            console.log("Auth state:", { isAuthenticated, role: auth.role, email: auth.email });
            fetchExtraHoursUsers();
        }

    }, [isAuthenticated, auth]);

    /*useEffect(() => {
        if (isAuthenticated && auth.role === 'USER' && auth.email) {
            fetchExtraHoursUsersByUser();
        } else if (isAuthenticated && auth.role === 'ADMIN') {
            fetchExtraHoursUsers();
        }
    }, [isAuthenticated, auth, fetchExtraHoursUsersByUser, fetchExtraHoursUsers]);*/

    const searchUser = useCallback(async (employeeId) => {

        try {
            setLoading(true);
            const userData = await UserService.searchUser(auth.token, employeeId);
            setError(null);
            return userData;
        } catch (err) {
            console.error('Error consultado el empleado:', err);
            setError('Error consultado el empleado:' + err.message);
            return null;
        } finally {
            setLoading(false);
        }
    }, [auth]);

    const handleCreate = async (newData) => {
        try {
            const newHour = await UserService.createExtraHour(auth.token, newData);
            setExtraHours([...extraHours, newHour]);
            setCreateModalOpen(false);
        } catch (err) {
            console.error('Error creating extra hour:', err);
            setError('Error creating extra hour');
        }
    };

    const handleUpdate = async (id, updatedData) => {
        try {
            await UserService.updateExtraHour(auth.token, id, updatedData);
            setExtraHours(
                extraHours.map((hour) =>
                    hour.id === id ? { ...hour, ...updatedData } : hour
                )
            );
            setUpdateModalOpen(false);
        } catch (err) {
            console.error('Error updating extra hour:', err);
            setError('Error updating extra hour');
        }
    };

    const handleDelete = async (id) => {
        try {
            await UserService.deleteExtraHour(auth.token, id);
            setExtraHours(extraHours.filter((hour) => hour.id !== id));
            setDeleteModalOpen(false);
        } catch (err) {
            console.error('Error deleting extra hour:', err);
            setError('Error deleting extra hour');
        }
    };

    if (loading) {
        return <div className="flex justify-center items-center p-4">Loading...</div>;
    }

    if (error) {
        return <div className="text-red-500 p-4">{error}</div>;
    }

    return (
        <div className="table-container">
            <h2>Horas Extra Amadeus</h2>
            <button
                className="update-button"
                onClick={() => setCreateModalOpen(true)}
            >
                Crear nuevo registro
            </button>
            <button
                className="update-button"
                onClick={() => setCreateModalOpen(true)}
            >
                Modal Prueba
            </button>
            {extraHours.length > 0 ? (
                <div className="overflow-x-auto">
                    <center>
                        <table>
                            <thead>
                                <tr>
                                    <th>ID</th>
                                    <th>Id empleado</th>
                                    <th>Nombre Empleado</th>
                                    <th>Cargo</th>
                                    <th>Salario</th>
                                    <th>Area</th>
                                    <th>% H/Extra</th>
                                    <th>Desc H/Extra</th>
                                    <th>startDatetime</th>
                                    <th>endDatetime</th>
                                    <th>hourPrice</th>
                                    <th>amountExtraHours</th>
                                    <th>comments</th>
                                    <th>totalExtraHour</th>
                                    <th>totalPayment</th>
                                </tr>
                            </thead>
                            <tbody>
                                {extraHours.map((hour) => (
                                    <tr key={hour.id} className="hover:bg-gray-50">
                                        <td>{hour.id}</td>
                                        <td>{hour.employee.employeeId}</td>
                                        <td >{hour.employee.employeeName}</td>
                                        <td>{hour.employee.job.jobName}</td>
                                        <td>{hour.employee.salary}</td>
                                        <td>{hour.employee.area.areaName}</td>
                                        <td>{hour.extraHourType.percentage}</td>
                                        <td>{hour.extraHourType.percentage}</td>
                                        <td>{hour.startDatetime}</td>
                                        <td>{hour.endDatetime}</td>
                                        <td>{hour.hourPrice}</td>
                                        <td>{hour.amountExtraHours}</td>
                                        <td>{hour.comments}</td>
                                        <td>{hour.totalExtraHour}</td>
                                        <td>{hour.totalPayment}</td>
                                        <td>

                                            <button className="update-button">
                                                <Link to={`/update-user/${hour.id}`}>
                                                    Update
                                                </Link>
                                                <button
                                                    className="update-button"
                                                    onClick={() => {
                                                        setSelectedHour(hour);
                                                        setUpdateModalOpen(true);
                                                    }}
                                                >
                                                    Update
                                                </button>
                                            </button>
                                            <button className="delete-button">Delete</button>
                                        </td>
                                    </tr>
                                ))}
                            </tbody>
                        </table>
                    </center>
                </div>
            ) : (
                <div className="text-center py-4 text-gray-500">
                    No se encontraron registros de horas extra
                </div>
            )}

            <CreateModal
                isOpen={createModalOpen}
                onClose={() => setCreateModalOpen(false)}
                onCreate={handleCreate}
                onSearch={searchUser}
            />
            {/*<EmployeeSearchModal
                isOpen={createModalOpen}
                onClose={() => setCreateModalOpen(false)}
                //onCreate={handleCreate}
                onSearch={searchUser}
            />*/}
            <UpdateModal
                isOpen={updateModalOpen}
                onClose={() => setUpdateModalOpen(false)}
                onUpdate={handleUpdate}
                hour={selectedHour}
            />
            <DeleteModal
                isOpen={deleteModalOpen}
                onClose={() => setDeleteModalOpen(false)}
                onDelete={handleDelete}
                id={selectedHour?.id}
            />

        </div>
    );
};

export default UsersExtraHours;`

Modal de crear, donde consulta un empleado, trae algunos campos y deja otros campos para diligenciarlos: 

`import React, { useState, useEffect } from 'react';
//import { Dialog, DialogTitle, DialogContent, DialogActions, Button, Input, Select } from '@/components/ui/dialog';
import Dialog from '@mui/material/Dialog';
import DialogTitle from '@mui/material/DialogTitle';
import DialogContent from '@mui/material/DialogContent';
import DialogActions from '@mui/material/DialogActions';
import Button from '@mui/material/Button';
import Input from '@mui/material/Input';
//import Select from '@mui/material/Select';

const CreateModal = ({ isOpen, onClose, onSearch, onCreate }) => {

    //const [employeeId, setEmployeeId] = useState('');

    const [formData, setFormData] = useState({
        employeeId: '',
        employeeName: '',
        jobName: '',
        salary: '',
        areaName: '',
        percentage: '',
        description: '',
        startDatetime: '',
        endDatetime: '',
        hourPrice: '',
        amountExtraHours: '',
        comments: '',
        totalExtraHour: '',
        totalPayment: '',
    });

    // Add debugging useEffect
    /*useEffect(() => {
        console.log("Form Data Updated:", formData);
    }, [formData]);*/

    const handleInputChange = (e) => {
        //setFormData({ ...formData, [e.target.name]: e.target.value });
        //setFormData((prevFormData) => ({ ...prevFormData, [e.target.name]: e.target.value }));
        const { name, value } = e.target;
        setFormData(prevState => ({
            ...prevState,
            [name]: value
        }));
    };


    const handleSearch = async () => {
        if (formData.employeeId) {
            try {
                const result = await onSearch(formData.employeeId);
                console.log("RESPUESTA RESULTADO MODAL: ", result.employeeId, result.salary);

                if (result) {
                    setFormData({
                        //...prev,
                        ...formData,
                        employeeName: result.employeeName || '',
                        jobName: result.jobName || '',
                        salary: result.salary?.toString() || '',
                        areaName: result.areaName || '',
                        hourPrice: result.hourPrice?.toString() || '',
                        // Map other fields as needed
                    });
                    console.log("RESPUESTA 2 RESULTADO MODAL: ", result.employeeName, result.jobName, result.areaName, result.employeeId, result.salary);
                    console.log("RESPUESTA 3 RESULTADO MODAL: ", formData.employeeName, formData.jobName, formData.salary, formData.areaName, formData.hourPrice);
                } else {
                    alert('Empleado no encontrado');
                }
            } catch (error) {
                console.error("Error en búsqueda:", error);
                alert('Error al buscar empleado');
            }
        }
    };

    /*const handleSearch = async () => {
        if (formData.employeeId) {
            try {
                const result = await onSearch(formData.employeeId);

                if (result) {
                    setFormData({
                        ...formData,
                        employeeName: result.employeeName || '',
                        jobName: result.jobName || '',
                        salary: result.salary?.toString() || '',
                        areaName: result.areaName || '',
                        hourPrice: result.hourPrice?.toString() || '',
                    });
                } else {
                    alert('Empleado no encontrado');
                }
            } catch (error) {
                console.error("Error en búsqueda:", error);
                alert('Error al buscar empleado');
            }
        }
    };*/

    /*const handleSearch = async () => {
        if (formData.employeeId) {
            try {
                const result = await onSearch(formData.employeeId);

                if (result) {
                    console.log("resultado búsqueda dentro del id:", result);
                    setFormData(prevState => ({
                        ...prevState,
                        employeeName: result.employeeName || '',
                        jobName: result.jobName || '',
                        salary: result.salary?.toString() || '',
                        areaName: result.areaName || '',
                        hourPrice: result.hourPrice?.toString() || '',
                    }));
                } else {
                    alert('Empleado no encontrado');
                }
            } catch (error) {
                console.error("Error en búsqueda:", error);
                alert('Error al buscar empleado');
            }
        }
    };*/

    /*const handleSearch = async () => {
        //if (formData.employeeId) {
        if (employeeId) {
            try {
                //const result = await onSearch(formData.employeeId);
                const result = await onSearch(employeeId);
                console.log("API Search Result:", result);

                if (result) {
                    // Directly set formData fields to values from result
                    setFormData({
                        employeeId: result.employeeId || '',
                        employeeName: result.employeeName || '',
                        jobName: result.jobName || '',
                        salary: result.salary?.toString() || '',
                        areaName: result.areaName || '',
                        hourPrice: result.hourPrice?.toString() || '',
                        /*percentage: formData.percentage,
                        description: formData.description,
                        startDatetime: formData.startDatetime,
                        endDatetime: formData.endDatetime,
                        amountExtraHours: formData.amountExtraHours,
                        comments: formData.comments,
                        totalExtraHour: formData.totalExtraHour,
                        totalPayment: formData.totalPayment,
                        });
                    console.log("Updated Form Data:", formData);
                } else {
                alert('Empleado no encontrado');
                }
                } catch (error) {
            console.error("Error en búsqueda:", error);
             alert('Error al buscar empleado');
            }
        }
    };*/


    const handleSubmit = () => {
        onCreate(formData);
        onClose();
    };

    return (
        <Dialog open={isOpen} onClose={onClose}
            classes={{ paper: 'customDialog' }} // Customize width and height here
        >
            <DialogTitle>Crear nuevo registro</DialogTitle>
            <DialogContent>
                <div className="grid grid-cols-2 gap-4">
                    <p>ID</p>{/*<Input name="employeeId" label="ID empleado" value={formData.employeeId} onChange={handleInputChange} />*/}
                    <div className="flex">
                        <Input
                            name="employeeId"
                            //value={formData.employeeId}
                            value={formData.employeeId}
                            onChange={handleInputChange}
                            //onChange={(e) => setEmployeeId(e.target.value)}
                            placeholder="ID de Empleado"
                        />
                        <Button onClick={handleSearch}>Buscar</Button>
                    </div>

                    <div className="col-span-2">
                        <h6>Nombre Empleado: <span className="text-red-500">{formData.employeeName}</span></h6>
                        <h6>Cargo: <span className="text-red-500">{formData.jobName}</span></h6>
                        <h6>Salario: <span className="text-red-500">{formData.salary}</span></h6>
                        <h6>Area: <span className="text-red-500">{formData.areaName}</span></h6>
                        <h6>Hour Price: <span className="text-red-500">{formData.hourPrice}</span></h6>
                    </div>
                    {/*<h6>Job Name: <p style={{ color: 'red' }}> {formData.jobName}</p></h6>
                    <p>Nombre Empleado</p><Input name="employeeName" label="Nombre empleado" value={formData.employeeName} onChange={handleInputChange} />
                    <p>Cargo</p><Input name="jobName" label="Cargo" value={formData.jobName} onChange={handleInputChange} />
                    <p>Salario</p><Input name="salary" label="Salario" value={formData.salary} onChange={handleInputChange} />
                    <p>Area</p><Input name="areaName" label="Area" value={formData.areaName} onChange={handleInputChange} />
                    <p>hourPrice</p><Input name="hourPrice" label="hourPrice" value={formData.hourPrice} onChange={handleInputChange} />*/}
                    {/*<p>% H/Extra</p><Input name="percentage" label="% H/Extra" value={formData.percentage} />
                    <p>Desc H/Extra</p><Input name="description" label="Desc H/Extra" value={formData.description} onChange={handleInputChange} />
                    <p>startDatetime</p><Input name="startDatetime" label="startDatetime" value={formData.startDatetime} onChange={handleInputChange} />
                    <p>endDatetime</p><Input name="endDatetime" label="endDatetime" value={formData.endDatetime} onChange={handleInputChange} />
                    <p>amountExtraHours</p><Input name="amountExtraHours" label="amountExtraHours" value={formData.amountExtraHours} onChange={handleInputChange} />
                    <p>comments</p><Input name="comments" label="comments" value={formData.comments} onChange={handleInputChange} />
                    <p>totalExtraHour</p><Input name="totalExtraHour" label="totalExtraHour" value={formData.totalExtraHour} onChange={handleInputChange} />
                    <p>totalPayment</p><Input name="totalPayment" label="totalPayment" value={formData.totalPayment} onChange={handleInputChange} />*/}
                </div>
            </DialogContent>
            <DialogActions>
                <Button variant="secondary" onClick={onClose}>
                    Cancelar
                </Button>
                <Button variant="primary" onClick={handleSubmit}>
                    Crear
                </Button>
            </DialogActions>
        </Dialog>
    );
};

export default CreateModal;`

Quedo atento a sus comentarios; cordialmente: @kanjimon1
